packages/llmRunner/src/llmRunner/LLMStep.ts

```diff
@@ -0,0 +1,235 @@
+import { SqError, SqProject } from "@quri/squiggle-lang";
+
+import { Artifact, ArtifactKind } from "./Artifact";
+import { LlmMetrics, Message } from "./LLMClient";
+import { LogEntry, Logger, TimestampedLogEntry } from "./Logger";
+import { PromptPair } from "./prompts";
+import { Workflow } from "./Workflow";
+
+export type CodeState =
+  | {
+      type: "formattingFailed";
+      error: string;
+      code: string;
+    }
+  | { type: "runFailed"; code: string; error: SqError; project: SqProject }
+  | { type: "success"; code: string };
+
+export function codeStateErrorString(codeState: CodeState): string {
+  if (codeState.type === "formattingFailed") {
+    return codeState.error;
+  } else if (codeState.type === "runFailed") {
+    return codeState.error.toStringWithDetails();
+  }
+  return "";
+}
+
+export type StepShape<
+  I extends Record<string, ArtifactKind> = Record<string, ArtifactKind>,
+  O extends Record<string, ArtifactKind> = Record<string, ArtifactKind>,
+> = {
+  inputs: I;
+  outputs: O;
+};
+
+type ExecuteContext<Shape extends StepShape> = {
+  getInput<K extends keyof Shape["inputs"]>(key: K): Inputs<Shape>[K];
+  setOutput<K extends keyof Shape["outputs"]>(
+    key: K,
+    value: Outputs<Shape>[K]
+  ): void;
+  queryLLM(promptPair: PromptPair): Promise<string | null>;
+  log(log: LogEntry): void;
+  workflow: Workflow; // TODO - shouldn't be exposed
```

Comment:
This is temporary until we get controllers.

Right now steps decide which steps to inject to the workflow; in the future the control loop should be responsible for that. (Probably... I'm not 100% sure yet, might change my mind later. Injecting steps from steps is convenient but less flexible; injecting steps from the controller is more flexible but requires either maintaining some kind of state in the controller, or assigning unique tags to steps. So it's a non-trivial trade-off.)